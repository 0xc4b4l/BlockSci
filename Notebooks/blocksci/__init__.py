from blocksci.blocksci_interface import *
from multiprocess import Pool
import psutil
import tempfile
import importlib
import subprocess
import sys
import os
import inspect

def map_blocks_general(chain, start, end, mapFunc):
    cpu_count = psutil.cpu_count()
    segments = chain.segment(start, end, cpu_count)
    with Pool(cpu_count - 1) as p:
        results_future = p.map_async(mapFunc, segments[1:])
        last = mapFunc(segments[0])
        results = results_future.get()
        results.insert(0, last)
    return [x for y in results for x in y]

def map_blocks(self, start, end, blockFunc):
    def mapFunc(blocks):
        return [blockFunc(block) for block in blocks]
    
    return map_blocks_general(self, start, end, mapFunc)

def filter_txes(self, start, end, filterFunc):
    def mapFunc(blocks):
        return [tx for block in blocks for tx in block if filterFunc(tx)]
    return map_blocks_general(self, start, end, mapFunc)

old_init = Blockchain.__init__
def new_init(self, loc = "/data2/bitcoin-data/"):
    old_init(self, loc)

Blockchain.__init__ = new_init
Blockchain.map_blocks = map_blocks
Blockchain.filter_txes = filter_txes
  
moduleDirectory = tempfile.TemporaryDirectory()
sys.path.append(moduleDirectory.name)
print(moduleDirectory.name)
dynamicFunctionCounter = 0

class DummyClass:
    pass 

loaderDirectory = os.path.dirname(os.path.abspath(inspect.getsourcefile(DummyClass)))

def function_loader(code, function_name):
    global dynamicFunctionCounter
    global moduleDirectory
    def generate_code(func_name, module_name, func_def):
        from string import Template
        filein = open(loaderDirectory + '/templateExtension.cpp')
        template = Template(filein.read())
        return template.safe_substitute({"func_name" : func_name, "module_name":module_name, "func_def" : func_def})
        
    def create_makefile(module_name):
        from string import Template
        import os
        filein = open(loaderDirectory + '/templateMakefile')
        template = Template(filein.read())
        subs = {"module_name" : module_name, "install_location" : moduleDirectory.name, "srcname" : module_name + ".cpp", "loaderDirectory":loaderDirectory}
        return template.safe_substitute(subs)
    module_name = "dynamicCode" + str(dynamicFunctionCounter)
    dynamicFunctionCounter += 1
    full_code = generate_code(function_name, module_name, code)
    makefile = create_makefile(module_name)
    builddir = tempfile.TemporaryDirectory()
    print(builddir)
    with open(builddir.name + '/' + module_name + ".cpp", 'w') as f:
        f.write(full_code)
    with open(builddir.name + '/CMakeLists.txt', 'w') as f:
        f.write(makefile)
    process = subprocess.Popen(["cmake", "."], cwd=builddir.name, stdout=subprocess.PIPE, stderr=subprocess.PIPE)
    out, err = process.communicate()
    print(err.decode('utf8'))
    process = subprocess.Popen(["make"], cwd=builddir.name, stdout=subprocess.PIPE, stderr=subprocess.PIPE)
    out, err = process.communicate()
    print(err.decode('utf8'))
    process = subprocess.Popen(["make", "install"], cwd=builddir.name, stdout=subprocess.PIPE, stderr=subprocess.PIPE)
    out, err = process.communicate()
    print(err.decode('utf8'))
    mod = importlib.import_module(module_name)
    return getattr(mod, "func")    
