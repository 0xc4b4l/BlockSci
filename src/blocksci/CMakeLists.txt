set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -march=native -mtune=native")

find_package( Boost 1.58 COMPONENTS filesystem iostreams serialization REQUIRED )
find_package(OpenSSL REQUIRED)

set(INCLUDE_INSTALL_DIR include/blocksci )
set(LIB_INSTALL_DIR lib/ )
set(SYSCONFIG_INSTALL_DIR etc/blocksci/ . )

set(BLOCKSCI_HEADERS
  address.hpp
  blocksci_fwd.hpp
  blocksci.hpp
  chain.hpp
  heuristics.hpp
  script.hpp
  cluster.hpp
  index.hpp
  exception.hpp
)

set(BLOCKSCI_SOURCES
  exception.cpp
)

set(ADDRESS_HEADERS
  address/address_fwd.hpp
  address/address_info.hpp
  address/address_types.hpp
  address/address.hpp
  address/raw_address.hpp
  address/dedup_address_info.hpp
  address/dedup_address_type.hpp
  address/dedup_address.hpp
  address/equiv_address_type.hpp
  address/equiv_address.hpp
)

set(ADDRESS_SOURCES
  address/address_info.cpp
  address/address.cpp
  address/raw_address.cpp
  address/dedup_address_info.cpp
  address/equiv_address.cpp
)

set(CHAIN_HEADERS
  chain/chain_fwd.hpp
  chain/algorithms.hpp
  chain/block.hpp
  chain/blockchain.hpp
  chain/chain_access.hpp
  chain/inout_pointer.hpp
  chain/inout.hpp
  chain/input.hpp
  chain/output.hpp
  chain/raw_block.hpp
  chain/raw_transaction.hpp
  chain/transaction_range.hpp
  chain/transaction_summary.hpp
  chain/transaction.hpp
)

set(CHAIN_SOURCES
  chain/block.cpp
  chain/blockchain.cpp
  chain/inout_pointer.cpp
  chain/input.cpp
  chain/output.cpp
  chain/transaction.cpp
)

set(HEURISTICS_HEADERS
  heuristics/blockchain_heuristics.hpp
  heuristics/change_address.hpp
  heuristics/taint.hpp
  heuristics/tx_identification.hpp
)

set(HEURISTICS_SOURCES
  heuristics/blockchain_heuristics.cpp
  heuristics/change_address.cpp
  heuristics/taint.cpp
  heuristics/tx_identification.cpp
)

set(INDEX_HEADERS
  index/address_index.hpp
  index/hash_index.hpp
  index/mempool_index.hpp
  index/column_iterator.hpp
)

set(INDEX_PRIVATE_HEADERS
  index/address_index_priv.hpp
  index/hash_index_priv.hpp
)

set(INDEX_SOURCES
  index/address_index.cpp
  index/address_index_priv.cpp
  index/hash_index.cpp
  index/hash_index_priv.cpp
  index/column_iterator.cpp
)

set(SCRIPT_HEADERS
  scripts/scripts_fwd.hpp
  scripts/bitcoin_base58.hpp
  scripts/bitcoin_bech32.hpp
  scripts/bitcoin_pubkey.hpp
  scripts/bitcoin_script.hpp
  scripts/bitcoin_segwit_addr.hpp
  scripts/multisig_script.hpp
  scripts/multisig_pubkey_script.hpp
  scripts/nonstandard_script.hpp
  scripts/nulldata_script.hpp
  scripts/pubkey_script.hpp
  scripts/pubkey_base_script.hpp
  scripts/script_access.hpp
  scripts/script_data.hpp
  scripts/script_info.hpp
  scripts/script_variant.hpp
  scripts/script_view.hpp
  scripts/script.hpp
  scripts/scripthash_script.hpp
)

set(SCRIPT_SOURCES
  scripts/bitcoin_base58.cpp
  scripts/bitcoin_bech32.cpp
  scripts/bitcoin_pubkey.cpp
  scripts/bitcoin_script.cpp
  scripts/bitcoin_segwit_addr.cpp
  scripts/script_view.cpp
  scripts/script.cpp
  scripts/script_variant.cpp
  scripts/scripthash_script.cpp
)

set_source_files_properties(scripts/bitcoin_bech32.cpp PROPERTIES COMPILE_FLAGS -Wno-everything)

set(UTIL_HEADERS
  util/file_mapper_fwd.hpp
  util/bitcoin_uint256.hpp
  util/data_access.hpp
  util/data_configuration.hpp
  util/file_mapper.hpp
  util/file_writer.hpp
  util/hash.hpp
  util/parallel.hpp
  util/progress_bar.hpp
  util/state.hpp
  util/util.hpp
  util/in_place_array.hpp
  util/apply_template.hpp
  util/dynamic_table.hpp
  util/static_table.hpp
  util/memory_view.hpp
)

set(UTIL_SOURCES
  util/bitcoin_uint256.cpp
  util/data_configuration.cpp
  util/hash.cpp
)

set(CLUSTER_HEADERS
  cluster/cluster_fwd.hpp
  cluster/cluster_manager.hpp
  cluster/cluster.hpp
)

set_source_files_properties(cluster/cluster_manager.cpp PROPERTIES COMPILE_FLAGS "-Wno-reserved-id-macro -Wno-shorten-64-to-32")

set(CLUSTER_SOURCES
  cluster/cluster_manager.cpp
  cluster/cluster.cpp
)

set(ALL_SOURCES ${BLOCKSCI_HEADERS} ${BLOCKSCI_SOURCES} ${ADDRESS_HEADERS} ${ADDRESS_SOURCES}  ${SCRIPT_HEADERS} ${SCRIPT_SOURCES} ${CHAIN_HEADERS} ${CHAIN_SOURCES} ${INDEX_HEADERS} ${INDEX_PRIVATE_HEADERS} ${INDEX_SOURCES} ${UTIL_HEADERS} ${UTIL_SOURCES} ${HEURISTICS_HEADERS} ${HEURISTICS_SOURCES} ${CLUSTER_HEADERS} ${CLUSTER_SOURCES})

add_library(blocksci SHARED ${ALL_SOURCES})
add_library(blocksci_static STATIC ${ALL_SOURCES})

include(GenerateExportHeader)
generate_export_header(blocksci)
set_target_properties(blocksci_static PROPERTIES COMPILE_FLAGS -DBLOCKSCI_STATIC_DEFINE)

set_property(TARGET blocksci PROPERTY VERSION ${Upstream_VERSION})
set_property(TARGET blocksci_static PROPERTY VERSION ${Upstream_VERSION})
set_property(TARGET blocksci_static PROPERTY POSITION_INDEPENDENT_CODE ON)

if( supported )
    message(STATUS "IPO / LTO enabled")
    set_property(TARGET blocksci PROPERTY INTERPROCEDURAL_OPTIMIZATION TRUE)
else()
    message(STATUS "IPO / LTO not supported: <${error}>")
endif()

source_group(blocksci\\chain FILES ${CHAIN_HEADERS} ${CHAIN_SOURCES})
source_group(blocksci\\address FILES ${ADDRESS_HEADERS} ${ADDRESS_SOURCES})
source_group(blocksci\\scripts FILES ${SCRIPT_HEADERS} ${SCRIPT_SOURCES})
source_group(blocksci\\index FILES ${INDEX_HEADERS} ${INDEX_SOURCES})
source_group(blocksci\\util FILES ${UTIL_HEADERS} ${UTIL_SOURCES})
source_group(blocksci\\heuristics FILES ${HEURISTICS_HEADERS} ${HEURISTICS_SOURCES})
source_group(blocksci\\cluster FILES ${CLUSTER_HEADERS} ${CLUSTER_SOURCES})
source_group(blocksci FILES ${BLOCKSCI_HEADERS} ${BLOCKSCI_SOURCES})

target_link_libraries( blocksci PUBLIC Threads::Threads)
target_link_libraries( blocksci PUBLIC Boost::filesystem Boost::iostreams Boost::serialization)
target_link_libraries( blocksci PRIVATE OpenSSL::Crypto )
target_link_libraries( blocksci PRIVATE rocksdb )
target_link_libraries( blocksci PRIVATE secp256k1)
target_link_libraries( blocksci PRIVATE dset)
target_link_libraries( blocksci PUBLIC blocksci_mpark_variant blocksci_ranges)

SET_TARGET_PROPERTIES(blocksci_static PROPERTIES OUTPUT_NAME blocksci)

target_link_libraries( blocksci_static PUBLIC Threads::Threads)
target_link_libraries( blocksci_static PUBLIC Boost::filesystem Boost::iostreams Boost::serialization)
target_link_libraries( blocksci_static PRIVATE OpenSSL::Crypto )
target_link_libraries( blocksci_static PRIVATE rocksdb )
target_link_libraries( blocksci_static PRIVATE secp256k1)
target_link_libraries( blocksci_static PRIVATE dset)
target_link_libraries( blocksci_static PUBLIC blocksci_mpark_variant blocksci_ranges)

target_include_directories(blocksci PUBLIC
  $<BUILD_INTERFACE:${CMAKE_CURRENT_SOURCE_DIR}/..>
  $<BUILD_INTERFACE:${CMAKE_CURRENT_BINARY_DIR}/..>
  $<INSTALL_INTERFACE:${INCLUDE_INSTALL_DIR}>
)

target_include_directories(blocksci_static PUBLIC
  $<BUILD_INTERFACE:${CMAKE_CURRENT_SOURCE_DIR}/..>
  $<BUILD_INTERFACE:${CMAKE_CURRENT_BINARY_DIR}/..>
  $<INSTALL_INTERFACE:${INCLUDE_INSTALL_DIR}>
)

SET(VERSION_STRING ${MAJOR_VERSION}.${MINOR_VERSION}.${PATCH_VERSION})

SET_TARGET_PROPERTIES(blocksci blocksci_static PROPERTIES
    VERSION "${VERSION_STRING}"
    SOVERSION "${SO_VERSION}")

include(CMakePackageConfigHelpers)

INSTALL(TARGETS blocksci EXPORT blocksci-targets LIBRARY DESTINATION ${LIB_INSTALL_DIR} ARCHIVE DESTINATION ${LIB_INSTALL_DIR})
INSTALL(TARGETS blocksci_static EXPORT blocksci_static LIBRARY DESTINATION ${LIB_INSTALL_DIR} ARCHIVE DESTINATION ${LIB_INSTALL_DIR})

export(EXPORT blocksci-targets
  FILE "${CMAKE_CURRENT_BINARY_DIR}/blocksci-targets.cmake"
)

set(ConfigPackageLocation "${LIB_INSTALL_DIR}/cmake/blocksci")

configure_package_config_file(blocksci-config.cmake
  ${CMAKE_CURRENT_BINARY_DIR}/blocksci-config.cmake
  INSTALL_DESTINATION "${ConfigPackageLocation}"
  PATH_VARS INCLUDE_INSTALL_DIR SYSCONFIG_INSTALL_DIR)

install (EXPORT blocksci-targets FILE blocksci-targets.cmake DESTINATION ${ConfigPackageLocation} )

install(
  FILES
    "${CMAKE_CURRENT_BINARY_DIR}/blocksci-config.cmake"
  DESTINATION
    ${ConfigPackageLocation}
  COMPONENT
    Devel
)

install(FILES ${BLOCKSCI_HEADERS} DESTINATION ${INCLUDE_INSTALL_DIR})
install(FILES ${ADDRESS_HEADERS} DESTINATION ${INCLUDE_INSTALL_DIR}/address)
install(FILES ${SCRIPT_HEADERS} DESTINATION ${INCLUDE_INSTALL_DIR}/scripts)
install(FILES ${UTIL_HEADERS} DESTINATION ${INCLUDE_INSTALL_DIR}/util)
install(FILES ${INDEX_HEADERS} DESTINATION ${INCLUDE_INSTALL_DIR}/index)
install(FILES ${CHAIN_HEADERS} DESTINATION ${INCLUDE_INSTALL_DIR}/chain)
install(FILES ${HEURISTICS_HEADERS} DESTINATION ${INCLUDE_INSTALL_DIR}/heuristics)
install(FILES ${CLUSTER_HEADERS} DESTINATION ${INCLUDE_INSTALL_DIR}/cluster)
