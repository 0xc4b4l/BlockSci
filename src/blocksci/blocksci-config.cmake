@PACKAGE_INIT@ 

include(CMakeFindDependencyMacro)

find_dependency(Boost 1.58 COMPONENTS system filesystem thread)
find_dependency(Threads)

include("${CMAKE_CURRENT_LIST_DIR}/blocksci-targets.cmake")
