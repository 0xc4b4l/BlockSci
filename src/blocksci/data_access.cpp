//
//  data_access.cpp
//  BlockReader
//
//  Created by Harry Kalodner on 1/18/17.
//  Copyright Â© 2017 Harry Kalodner. All rights reserved.
//

#include "data_access.hpp"
#include "data_configuration.hpp"

#include <blocksci/scripts/script_access.hpp>
#include <blocksci/address/address_index.hpp>
#include <blocksci/chain/chain_access.hpp>
#include <blocksci/hash_index.hpp>

namespace blocksci {
    
    DataAccess::DataAccess(const DataConfiguration &config_, bool errorOnReorg_, uint32_t blocksIgnored_) : config(config_), chain{std::make_unique<ChainAccess>(config, errorOnReorg_, blocksIgnored_)}, scripts{std::make_unique<ScriptAccess>(config)}, addressIndex{std::make_unique<AddressIndex>(config)}, hashIndex{std::make_unique<HashIndex>(config.hashIndexFilePath().native())} {}
    
    DataAccess &DataAccess::Instance(const DataConfiguration &config_, bool errorOnReorg, uint32_t blocksIgnored) {
        // Since it's a static variable, if the class has already been created,
        // It won't be created again.
        // And it **is** thread-safe in C++11.
        
        static DataAccess myInstance(config_, errorOnReorg, blocksIgnored);
        static DataConfiguration config = config_;
        
        if (!config_.isNull()) {
            config = config_;
            myInstance.chain->reload();
            myInstance.scripts->reload();
        }
        
        // Return a reference to our instance.
        return myInstance;
    }
    
    DataAccess &DataAccess::Instance() {
        DataConfiguration config {};
        return Instance(config, true, 0);
    }
}



