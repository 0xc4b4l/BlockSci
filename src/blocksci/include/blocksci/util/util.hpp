//
//  util.hpp
//  blocksci
//
//  Created by Harry Kalodner on 9/6/17.
//
//

#ifndef blocksci_util_util_h
#define blocksci_util_util_h

#include <tuple>
#include <utility>

namespace blocksci {
    namespace internal {
        template<typename A, template<typename...> class O>
        struct to_variadic;
        
        template<template<typename...> class V, template<typename...> class O, typename ...A>
        struct to_variadic<V<A...>, O> {
            using type = O<A...>;
        };
        
        template<template<class> class K, typename T>
        struct map_variadic;
        
        template <template<class> class K, typename... Types>
        struct map_variadic<K, std::tuple<Types...>> {
            using type = std::tuple<K<Types>...>;
        };
    }
    
    template<typename A, template<typename...> class O>
    using to_variadic_t = typename internal::to_variadic<A, O>::type;
}

#endif /* blocksci_util_util_h */
