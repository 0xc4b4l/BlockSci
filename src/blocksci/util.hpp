//
//  util.hpp
//  blocksci
//
//  Created by Harry Kalodner on 9/6/17.
//
//

#ifndef util_h
#define util_h

#include <tuple>
#include <utility>

namespace blocksci {
    namespace internal {
        
        template<int I, class Tuple, typename F> struct for_each_impl {
            static void for_each(Tuple& t, F f) {
                for_each_impl<I - 1, Tuple, F>::for_each(t, f);
                f(std::get<I>(t));
            }
        };
        template<class Tuple, typename F> struct for_each_impl<0, Tuple, F> {
            static void for_each(Tuple& t, F f) {
                f(std::get<0>(t));
            }
        };
        
        template <class F, size_t... Is>
        constexpr auto index_apply_impl(F f, std::index_sequence<Is...>) {
            return f(std::integral_constant<size_t, Is> {}...);
        }
        
        template <size_t N, class F>
        constexpr auto index_apply(F f) {
            return index_apply_impl(f, std::make_index_sequence<N>{});
        }
        
        template <typename EnumStruct>
        struct MakeDynamicTableImpl {
            template<template<typename EnumStruct::Enum> class Functor, std::size_t... Is>
            constexpr auto makeTable(std::index_sequence<Is...>) -> std::array<decltype(&Functor<EnumStruct::all[0]>::f), sizeof...(Is)> {
                return {{Functor<EnumStruct::all[Is]>::f...}};
            }
        };
        
        
        template <typename EnumStruct, class ...Args>
        struct MakeStaticTableImpl {
            template<template<typename EnumStruct::Enum> class Functor, std::size_t... Is>
            constexpr auto makeTable(std::index_sequence<Is...>, Args... args) -> std::array<decltype(Functor<EnumStruct::all[0]>::f(args...)), sizeof...(Is)> {
                return {{Functor<EnumStruct::all[Is]>::f(args...)...}};
            }
        };
    }
    
    template <typename EnumStruct, template<typename EnumStruct::Enum> class Functor, class ...Args>
    constexpr auto make_static_table(Args... args) {
        internal::MakeStaticTableImpl<EnumStruct, Args...> tableMaker;
        return tableMaker.template makeTable<Functor>(std::make_index_sequence<EnumStruct::all.size()>{}, args...);
    }
    
    template <typename EnumStruct, template<typename EnumStruct::Enum> class Functor>
    constexpr auto make_dynamic_table() {
        internal::MakeDynamicTableImpl<EnumStruct> tableMaker;
        return tableMaker.template makeTable<Functor>(std::make_index_sequence<EnumStruct::all.size()>{});
    }
    
    template<class Tuple, typename F>
    void for_each(Tuple& t, F f) {
        internal::for_each_impl<std::tuple_size<Tuple>::value - 1, Tuple, F>::for_each(t, f);
    }
    
    template <class Tuple, class F>
    constexpr auto apply(Tuple t, F f) {
        return internal::index_apply<std::tuple_size<Tuple>{}>([&](auto... Is) {
            return std::make_tuple(f(std::get<Is>(t))...);
        });
    }
}

#endif /* util_h */
