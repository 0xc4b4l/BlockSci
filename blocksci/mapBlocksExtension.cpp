#include <blocksci/blocksci.hpp>
#include <pybind11/pybind11.h>
#include <pybind11/stl.h>

using namespace blocksci;
namespace py = pybind11;

auto mapFunc(const Block &block) {
    return ${func_def};
}


PYBIND11_MODULE(${module_name}, m) {
    m.def("func", [](const Blockchain &chain, uint32_t start, uint32_t stop) {
    	return chain.map<decltype(mapFunc(std::declval<Block>()))>(start, stop, mapFunc);
    });
}