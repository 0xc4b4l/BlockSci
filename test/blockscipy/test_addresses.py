import blocksci
from util import Coin

FEE = 0.0001
ADDR_TYPES = ['p2pkh', 'p2sh', 'p2wpkh', 'p2wsh']


def addresses(chain, json_data):
    for addr_type in ADDR_TYPES:
        for i in range(3):
            addr = chain.address_from_string(json_data["address-{}-spend-{}".format(addr_type, i)])
            yield addr, addr_type


def address_received_test(addr, address_type, balance, ntxes):
    assert address_type == addr.type
    assert Coin(balance) == addr.balance()
    assert ntxes == len(addr.txes())


def test_p2pkh_address(chain, json_data):
    addr = chain.address_from_string(json_data['address-p2pkh-spend-0'])
    address_received_test(addr, blocksci.address_type.pubkeyhash, 1, 1)

    addr = chain.address_from_string(json_data['address-p2pkh-spend-1'])
    address_received_test(addr, blocksci.address_type.pubkeyhash, 0, 2)

    addr = chain.address_from_string(json_data['address-p2pkh-spend-2'])
    address_received_test(addr, blocksci.address_type.pubkeyhash, 0, 4)


def test_p2sh_address(chain, json_data):
    addr = chain.address_from_string(json_data['address-p2sh-spend-0'])
    address_received_test(addr, blocksci.address_type.scripthash, 1, 1)

    addr = chain.address_from_string(json_data['address-p2sh-spend-1'])
    address_received_test(addr, blocksci.address_type.scripthash, 0, 2)

    addr = chain.address_from_string(json_data['address-p2sh-spend-2'])
    address_received_test(addr, blocksci.address_type.scripthash, 0, 4)


def test_p2wpkh_address(chain, json_data):
    addr = chain.address_from_string(json_data['address-p2wpkh-spend-0'])
    address_received_test(addr, blocksci.address_type.witness_pubkeyhash, 1, 1)

    addr = chain.address_from_string(json_data['address-p2wpkh-spend-1'])
    address_received_test(addr, blocksci.address_type.witness_pubkeyhash, 0, 2)

    addr = chain.address_from_string(json_data['address-p2wpkh-spend-2'])
    address_received_test(addr, blocksci.address_type.witness_pubkeyhash, 0, 4)


def test_p2wsh_address(chain, json_data):
    addr = chain.address_from_string(json_data['address-p2wsh-spend-0'])
    address_received_test(addr, blocksci.address_type.witness_scripthash, 1, 1)

    addr = chain.address_from_string(json_data['address-p2wsh-spend-1'])
    address_received_test(addr, blocksci.address_type.witness_scripthash, 0, 2)

    addr = chain.address_from_string(json_data['address-p2wsh-spend-2'])
    address_received_test(addr, blocksci.address_type.witness_scripthash, 0, 4)


def test_address_regression(chain, json_data, regtest):
    for addr, addr_type in addresses(chain, json_data):
        print(addr.address_string, file=regtest)
        print(addr.balance(), file=regtest)
        print(addr.equiv(), file=regtest)
        print(addr.has_been_spent, file=regtest)
        print(addr.in_txes(), file=regtest)
        print(addr.in_txes_count(), file=regtest)
        print(addr.out_txes(), file=regtest)
        print(addr.out_txes_count(), file=regtest)
        if addr_type[-2:] != "sh":
            print(addr.pubkey, file=regtest)
            print(addr.pubkeyhash, file=regtest)
        print(addr.raw_type, file=regtest)
        print(addr.revealed_tx, file=regtest)
        print(addr.type, file=regtest)
