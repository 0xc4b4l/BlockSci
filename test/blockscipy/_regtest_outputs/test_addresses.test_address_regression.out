myowM8GghMuCX4tn58Nmh9eubwMjuwKiCJ
100000000
EquivAddress(PubkeyHashAddress(myowM8GghMuCX4tn58Nmh9eubwMjuwKiCJ))
False
[]
0
[Tx(len(txins)=1, len(txouts)=2, size_bytes=225, block_height=111, tx_index=112)]
1
None
7fd8abeed9bbe533e8bac0d38732dd9c7f8ba8c8
2
None
Pay to pubkey hash
mrPRMELJcGsbKmuFiGwUgNCX9Wo4iSy44T
0
EquivAddress(PubkeyHashAddress(mrPRMELJcGsbKmuFiGwUgNCX9Wo4iSy44T))
True
[Tx(len(txins)=1, len(txouts)=1, size_bytes=192, block_height=112, tx_index=115)]
1
[Tx(len(txins)=1, len(txouts)=2, size_bytes=226, block_height=112, tx_index=114)]
1
b'\x02\x9e\xc4{e\xef\xbd\xd2N\xbcp\x10\xfbg\x1cz\xab\x92\x8e\x91"4?9\rpq\xda\x15\xe5l\xf5!'
dd9f85e63f3e6f87bb9ee130bb2c8a62dea33c77
2
Tx(len(txins)=1, len(txouts)=1, size_bytes=192, block_height=112, tx_index=115)
Pay to pubkey hash
mk7GYZDvnyyTyi2anTr5w5sTqW3r18WQWb
0
EquivAddress(PubkeyHashAddress(mk7GYZDvnyyTyi2anTr5w5sTqW3r18WQWb))
True
[Tx(len(txins)=1, len(txouts)=1, size_bytes=192, block_height=113, tx_index=118), Tx(len(txins)=1, len(txouts)=1, size_bytes=192, block_height=113, tx_index=120)]
2
[Tx(len(txins)=1, len(txouts)=2, size_bytes=226, block_height=113, tx_index=117), Tx(len(txins)=1, len(txouts)=2, size_bytes=226, block_height=113, tx_index=119)]
2
b'\x03\x08\xde\xe8\xbb\xccv\xa7\xad\xdfm\xd7\xa2\xdam\xbf\xa8\xa0N(\x03\xd1\r\x98\x8c)\x8d\x12\x9f\xd3l\xae\xea'
9a337fd549963efeb8780c93e234bbf754e15d32
2
Tx(len(txins)=1, len(txouts)=1, size_bytes=192, block_height=113, tx_index=118)
Pay to pubkey hash
2Mshjw3i5VcfqCm6FaJLxLSdSQhShKJYk6d
100000000
EquivAddress(ScriptHashAddress(2Mshjw3i5VcfqCm6FaJLxLSdSQhShKJYk6d))
False
[]
0
[Tx(len(txins)=1, len(txouts)=2, size_bytes=223, block_height=114, tx_index=122)]
1
4
None
Pay to script hash
2Mzqim7c3TBkpRyqXmB23DXXcm1nHN6kFsf
0
EquivAddress(ScriptHashAddress(2Mzqim7c3TBkpRyqXmB23DXXcm1nHN6kFsf))
True
[Tx(len(txins)=1, len(txouts)=1, size_bytes=193, block_height=115, tx_index=125)]
1
[Tx(len(txins)=1, len(txouts)=2, size_bytes=224, block_height=115, tx_index=124)]
1
4
Tx(len(txins)=1, len(txouts)=1, size_bytes=193, block_height=115, tx_index=125)
Pay to script hash
2N1otuhuzEiwReoLqqrJKTk67A2aW5ZXXDW
0
EquivAddress(ScriptHashAddress(2N1otuhuzEiwReoLqqrJKTk67A2aW5ZXXDW))
True
[Tx(len(txins)=1, len(txouts)=1, size_bytes=193, block_height=116, tx_index=128), Tx(len(txins)=1, len(txouts)=1, size_bytes=194, block_height=116, tx_index=130)]
2
[Tx(len(txins)=1, len(txouts)=2, size_bytes=223, block_height=116, tx_index=127), Tx(len(txins)=1, len(txouts)=2, size_bytes=223, block_height=116, tx_index=129)]
2
4
Tx(len(txins)=1, len(txouts)=1, size_bytes=193, block_height=116, tx_index=128)
Pay to script hash
bcrt1qqlxxeyqj0ka4kqpzajv27l8px06neuqcwxg62e
100000000
EquivAddress(WitnessPubkeyAddress(bcrt1qqlxxeyqj0ka4kqpzajv27l8px06neuqcwxg62e))
False
[]
0
[Tx(len(txins)=1, len(txouts)=2, size_bytes=223, block_height=117, tx_index=132)]
1
None
18f03cf533e17caf98ec22005bbb7d12906ccc07
7
None
Pay to witness pubkey hash
bcrt1qxxhhtzpwdfnvckc5w6p89ulje744wfdjdv3k3m
0
EquivAddress(WitnessPubkeyAddress(bcrt1qxxhhtzpwdfnvckc5w6p89ulje744wfdjdv3k3m))
True
[Tx(len(txins)=1, len(txouts)=1, size_bytes=113, block_height=118, tx_index=135)]
1
[Tx(len(txins)=1, len(txouts)=2, size_bytes=222, block_height=118, tx_index=134)]
1
b'\x02\xe4b\x1f5g{\x1a!\xa9;\xf9\xf0-\xbd\x80\x1d\xfd\xaa\x0cbU[\x90\x16\x1a\xe2\xa2,\xe85\xc1\xd6'
b22557abcff2f3728276145bcc666a2e8875af31
7
Tx(len(txins)=1, len(txouts)=1, size_bytes=113, block_height=118, tx_index=135)
Pay to witness pubkey hash
bcrt1qzs9mywz92nu34gyflypumr9faw7swx2wdnnt6u
0
EquivAddress(WitnessPubkeyAddress(bcrt1qzs9mywz92nu34gyflypumr9faw7swx2wdnnt6u))
True
[Tx(len(txins)=1, len(txouts)=1, size_bytes=113, block_height=119, tx_index=138), Tx(len(txins)=1, len(txouts)=1, size_bytes=113, block_height=119, tx_index=140)]
2
[Tx(len(txins)=1, len(txouts)=2, size_bytes=222, block_height=119, tx_index=137), Tx(len(txins)=1, len(txouts)=2, size_bytes=223, block_height=119, tx_index=139)]
2
b'\x02\xf5\x96<\xb8\xdf8\xcd\x02"_X.!`\xc5\xac\xbc\xb9T\xfa\xe3@p\xb62\t\xd6\x84\xaf\xa5\x9d\xd1'
4e1907bdeba98ccd03f989a01af9544538b20b14
7
Tx(len(txins)=1, len(txouts)=1, size_bytes=113, block_height=119, tx_index=138)
Pay to witness pubkey hash
bcrt1qgzdvhl0u689maa4jtpfa2v7n9hshvfuph028588dp2lw2kuqlc3qlkc9dx
100000000
EquivAddress(WitnessScriptHashAddress(bcrt1qgzdvhl0u689maa4jtpfa2v7n9hshvfuph028588dp2lw2kuqlc3qlkc9dx))
False
[]
0
[Tx(len(txins)=1, len(txouts)=2, size_bytes=235, block_height=120, tx_index=142)]
1
8
None
Pay to witness script hash
bcrt1qektr66tljls59h4w5qwshr6fwr4qfeuwm3zkwcsd424sd046n92sagvgwa
0
EquivAddress(WitnessScriptHashAddress(bcrt1qektr66tljls59h4w5qwshr6fwr4qfeuwm3zkwcsd424sd046n92sagvgwa))
True
[Tx(len(txins)=1, len(txouts)=1, size_bytes=113, block_height=121, tx_index=145)]
1
[Tx(len(txins)=1, len(txouts)=2, size_bytes=235, block_height=121, tx_index=144)]
1
8
Tx(len(txins)=1, len(txouts)=1, size_bytes=113, block_height=121, tx_index=145)
Pay to witness script hash
bcrt1qe2stj85585lav85e4cgpcprxva6ghfdp978q4y0rjc8k9u0cc46q66sf9j
0
EquivAddress(WitnessScriptHashAddress(bcrt1qe2stj85585lav85e4cgpcprxva6ghfdp978q4y0rjc8k9u0cc46q66sf9j))
True
[Tx(len(txins)=1, len(txouts)=1, size_bytes=113, block_height=122, tx_index=148), Tx(len(txins)=1, len(txouts)=1, size_bytes=113, block_height=122, tx_index=150)]
2
[Tx(len(txins)=1, len(txouts)=2, size_bytes=235, block_height=122, tx_index=147), Tx(len(txins)=1, len(txouts)=2, size_bytes=235, block_height=122, tx_index=149)]
2
8
Tx(len(txins)=1, len(txouts)=1, size_bytes=113, block_height=122, tx_index=148)
Pay to witness script hash
